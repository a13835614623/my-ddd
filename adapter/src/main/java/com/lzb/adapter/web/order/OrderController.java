package com.lzb.adapter.web.order;

import com.lzb.app.order.cmd.PlaceOrderAppService;
import com.lzb.app.order.cmd.dto.PlaceOrderDto;
import com.lzb.component.dto.ResponseDto;
import lombok.RequiredArgsConstructor;

import org.springframework.context.annotation.Lazy;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/**
 * <br/>
 * Created on : 2023-09-06 13:34
 * @author lizebin
 */
@RestController
@RequestMapping("/order")
@RequiredArgsConstructor(onConstructor = @__(@Lazy))
public class OrderController {

    private final PlaceOrderAppService placeOrderAppService;

    @PutMapping
    public ResponseDto<Long> placeOrder(@RequestBody PlaceOrderDto placeOrderDto) {
        return ResponseDto.success(placeOrderAppService.placeOrder(placeOrderDto));
    }

}
