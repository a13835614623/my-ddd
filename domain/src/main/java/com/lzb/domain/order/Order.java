package com.lzb.domain.order;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Objects;

import cn.hutool.core.lang.UUID;
import com.lzb.domain.common.BaseAggregate;
import com.lzb.domain.order.event.OrderCanceledEvent;
import lombok.Builder;
import lombok.Getter;
import lombok.NonNull;
import lombok.experimental.SuperBuilder;
import lombok.extern.jackson.Jacksonized;
import lombok.extern.slf4j.Slf4j;

/**
 * 订单聚合根<br/>
 * 实体所有写的操作，必须通过聚合根调用，所以改成包可见(default)
 * 实体所有读操作，可以改成公共(public)
 * Created on : 2022-06-11 14:41
 *
 * @author lizebin
 */
@Slf4j
@Getter
// builder 反序列化
@SuperBuilder
@Jacksonized
public class Order extends BaseAggregate<Order> {

    @NonNull
    private String orderStatus;

    @NonNull
    private String currency;

    @NonNull
    private BigDecimal exchangeRate;

    @NonNull
    private BigDecimal totalShouldPay;

    @NonNull
    private BigDecimal totalActualPay;

    @NonNull
    private OrderAddress orderAddress;

    @Builder.Default
    private OrderDetails orderDetails = new OrderDetails(new ArrayList<>());

    public void updateTotalActualPay(BigDecimal totalActualPay) {
        this.totalActualPay = totalActualPay;
    }

    /**
     * 取消订单
     */
    public void cancel() {
        orderStatus = "CANCEL";
        orderDetails.forEach(OrderDetail::cancel);
        addEvent(OrderCanceledEvent.create(
                UUID.randomUUID().toString(),
                Objects.toString(id),
                Objects.toString(id),
                id));
    }
}
