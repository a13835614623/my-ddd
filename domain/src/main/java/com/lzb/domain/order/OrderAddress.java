package com.lzb.domain.order;

import com.lzb.domain.common.BaseEntity;
import com.lzb.domain.order.valobj.FullAddressLine;
import com.lzb.domain.order.valobj.FullName;
import lombok.Getter;
import lombok.experimental.SuperBuilder;

/**
 * 订单地址实体<br/>
 * Created on : 2023-08-31 12:52
 * @author lizebin
 */
@Getter
@SuperBuilder
public class OrderAddress extends BaseEntity<OrderAddress> {

    private Long orderId;

    private FullName fullName;

    private FullAddressLine fullAddressLine;

    /**
     * 邮箱
     */
    private String email;

    /**
     * 电话
     */
    private String phoneNumber;

    /**
     * 国家
     */
    private String country;

}
