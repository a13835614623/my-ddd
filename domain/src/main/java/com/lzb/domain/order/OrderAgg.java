package com.lzb.domain.order;

/**
 * 订单聚合根<br/>
 * Created on : 2023-08-30 22:35
 * @author mac
 */
public class OrderAgg {

}
