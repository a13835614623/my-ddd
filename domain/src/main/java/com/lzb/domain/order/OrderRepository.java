package com.lzb.domain.order;

import com.lzb.domain.common.repository.CommonRepository;

/**
 * <br/>
 * Created on : 2023-08-30 23:07
 * @author mac
 */
public interface OrderRepository extends CommonRepository<Order> {

}
