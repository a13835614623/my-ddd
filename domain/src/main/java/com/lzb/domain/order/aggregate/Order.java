package com.lzb.domain.order.aggregate;

import java.math.BigDecimal;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.lzb.component.exception.BizException;
import com.lzb.domain.common.aggregate.BaseAggregate;
import com.lzb.domain.order.aggregate.builder.OrderAddressBuilder;
import com.lzb.domain.order.dto.LockStockDto;
import com.lzb.domain.order.enums.OrderStatus;
import com.lzb.domain.order.event.OrderCanceledEvent;
import com.lzb.domain.order.event.OrderPlacedEvent;
import com.lzb.domain.order.valobj.FullAddressLine;
import com.lzb.domain.order.valobj.FullName;
import lombok.Getter;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;
import one.util.streamex.StreamEx;

/**
 * 订单聚合根<br/>
 * 实体所有写的操作，必须通过聚合根调用，所以改成包可见(default)
 * 实体所有读操作，可以改成公共(public)
 * Created on : 2022-06-11 14:41
 *
 * @author lizebin
 */
@Slf4j
@Getter
public class Order extends BaseAggregate<Order> {

    @NonNull
    private final OrderStatus orderStatus;

    @NonNull
    private final String currency;

    @NonNull
    private final BigDecimal exchangeRate;

    @NonNull
    private final BigDecimal totalShouldPay;

    @NonNull
    private final BigDecimal totalActualPay;

    @NonNull
    private final OrderAddress orderAddress;

    @NonNull
    private final OrderDetails orderDetails;


    /**
     * 仓储层使用，无需校验业务逻辑，直接构造
     * @param id
     * @param version
     * @param orderStatus
     * @param currency
     * @param exchangeRate
     * @param totalShouldPay
     * @param totalActualPay
     * @param orderAddress
     * @param orderDetails
     */
    public Order(long id,
            int version,
            @NonNull OrderStatus orderStatus,
            @NonNull String currency,
            @NonNull BigDecimal exchangeRate,
            @NonNull BigDecimal totalShouldPay,
            @NonNull BigDecimal totalActualPay,
            @NonNull OrderAddress orderAddress,
            @NonNull List<OrderDetail> orderDetails) {
        super(id);
        this.version = version;
        this.orderStatus = orderStatus;
        this.currency = currency;
        this.exchangeRate = exchangeRate;
        this.totalShouldPay = totalShouldPay;
        this.totalActualPay = totalActualPay;
        this.orderAddress = orderAddress;
        this.orderDetails = new OrderDetails(orderDetails);
    }


    public void updateAddress(String email,
            String phoneNumber, FullName fullName,
            FullAddressLine fullAddressLine, String country) {
        if (isShipped()) {
            throw new BizException("订单已发货，不能修改地址");
        }
        OrderAddress newOrderAddress = OrderAddressBuilder.newInstance(this.orderAddress)
                .addressLine1(fullAddressLine.getAddressLine1())
                .addressLine2(fullAddressLine.getAddressLine2())
                .country(country)
                .email(email)
                .firstName(fullName.getFirstName())
                .lastName(fullName.getLastName())
                .phoneNumber(phoneNumber).build();
        this.orderAddress = newOrderAddress;
    }

    /**
     * 更新库存结果
     * @param lockStock
     */
    public void updateLockStock(LockStockDto lockStock) {
        Map<Integer, List<OrderDetail>> skuId2OrderDetails = StreamEx.of(orderDetails.toStream()).groupingBy(OrderDetail::getSkuId);
        skuId2OrderDetails.keySet().forEach(skuId -> {
            List<OrderDetail> skuOrderDetails = skuId2OrderDetails.getOrDefault(skuId, Collections.emptyList());
            for (int i = 1; i <= skuOrderDetails.size(); i++) {
                skuOrderDetails.get(i - 1).updateLocked(i <= lockStock.getDetail(skuId).getLockedNum());
            }
        });
    }

    public void updateTotalActualPay(BigDecimal totalActualPay) {
        this.totalActualPay = totalActualPay;
    }

    /**
     * 取消订单
     */
    public void cancel() {
        orderStatus = OrderStatus.CANCELED;
        orderDetails.forEach(OrderDetail::cancel);
        addEvent(OrderCanceledEvent.create(id));
    }

    /**
     * 是否取消
     * @return
     */
    public boolean isCancel() {
        return OrderStatus.CANCELED == orderStatus;
    }

    /**
     * 生单
     */
    public void place() {
        addEvent(OrderPlacedEvent.create(this.id));
    }

    public boolean isShipped() {
        return false;
    }

    /**
     * 只更新姓名
     * @param fullName
     */
    public void updateFullName(FullName fullName) {
        this.orderAddress.updateFullName(fullName);
    }
}
