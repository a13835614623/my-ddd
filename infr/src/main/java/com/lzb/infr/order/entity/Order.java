package com.lzb.infr.order.entity;

import java.io.Serializable;
import java.math.BigDecimal;
import java.time.LocalDateTime;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableName;
import lombok.Data;
import lombok.EqualsAndHashCode;

/**
 * <p>
 * 
 * </p>
 *
 * @author lizebin
 * @since 2023-08-29
 */
@Data
@EqualsAndHashCode(callSuper = false)
@TableName("\"order\"")
public class Order implements Serializable {

    @TableId(type = IdType.AUTO)
    private Long orderId;

    private String orderStatus;

    private String currency;

    private BigDecimal exchangeRate;

    private BigDecimal totalShouldPay;

    private BigDecimal totalActualPay;

    private LocalDateTime addTime;

    private LocalDateTime updateTime;


}