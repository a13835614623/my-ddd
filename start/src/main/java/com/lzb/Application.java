package com.lzb;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.EnableAspectJAutoProxy;

@SpringBootApplication
// AopContext.currentProxy() 启用自身调用aop
@EnableAspectJAutoProxy(proxyTargetClass = true, exposeProxy = true)
public class Application {

	public static void main(String[] args) {
		SpringApplication.run(Application.class, args);
	}

}
