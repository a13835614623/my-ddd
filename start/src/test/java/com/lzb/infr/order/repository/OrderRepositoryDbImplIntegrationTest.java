package com.lzb.infr.order.repository;

import javax.annotation.Resource;

import com.lzb.BaseIntegrationTest;
import com.lzb.domain.order.Order;
import com.lzb.domain.order.OrderRepository;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import org.springframework.test.context.jdbc.Sql;

class OrderRepositoryDbImplIntegrationTest extends BaseIntegrationTest {

    @Resource(name = OrderRepositoryDbImpl.BEAN_NAME)
    private OrderRepository orderRepository;

    @Test
    @Sql("/sql/OrderRepositoryDbImplIntegrationTest/should_order_get.sql")
    @DisplayName("测试聚合根查询")
    void should_order_get() {
        long orderId = 1L;
        Order order = orderRepository.get(orderId).orElseThrow();
        assertJSON(order);
    }

}