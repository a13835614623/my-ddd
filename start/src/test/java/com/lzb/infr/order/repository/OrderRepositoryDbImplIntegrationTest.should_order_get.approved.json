{
  "id" : 1,
  "version" : 1,
  "orderStatus" : "WAIT_PAY",
  "currency" : "CNY",
  "exchangeRate" : 1,
  "totalShouldPay" : 88,
  "totalActualPay" : 88,
  "orderAddress" : {
    "id" : null,
    "version" : null,
    "orderId" : 1,
    "fullName" : {
      "firstName" : "first",
      "lastName" : "last"
    },
    "fullAddressLine" : {
      "addressLine1" : "line1",
      "addressLine2" : "line2"
    },
    "email" : "123456@qq.com",
    "phoneNumber" : "13800000000",
    "country" : "CHINA"
  },
  "orderDetails" : [ {
    "id" : 1,
    "version" : null,
    "orderId" : 1,
    "skuId" : 1,
    "orderStatus" : "WAIT_PAY",
    "price" : 1
  }, {
    "id" : 2,
    "version" : null,
    "orderId" : 1,
    "skuId" : 2,
    "orderStatus" : "WAIT_PAY",
    "price" : 10
  } ],
  "logs" : [ ],
  "events" : [ ]
}